contrib/twitchtv/twirp/twirp.go

```diff
@@ -0,0 +1,219 @@
+		tracer.Tag(ext.HTTPMethod, req.Method),
+		tracer.Tag(ext.HTTPURL, req.URL.Path),
+	}
+	if method, ok := twirp.MethodName(ctx); ok {
+		opts = append(opts, tracer.ResourceName(method))
+	}
+	if !math.IsNaN(wc.cfg.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, wc.cfg.analyticsRate))
+	}
+	if spanctx, err := tracer.Extract(tracer.HTTPHeadersCarrier(req.Header)); err == nil {
+		opts = append(opts, tracer.ChildOf(spanctx))
+	}
+
+	span, ctx := tracer.StartSpanFromContext(req.Context(), "twirp.request", opts...)
+	defer span.Finish()
+
+	req = req.WithContext(ctx)
+	err := tracer.Inject(span.Context(), tracer.HTTPHeadersCarrier(req.Header))
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "contrib/twitchtv/twirp.wrappedClient: failed to inject http headers: %v\n", err)
+	}
+
+	res, err := wc.c.Do(req)
+	if err != nil {
+		span.SetTag(ext.Error, err)
+	} else {
+		span.SetTag(ext.HTTPCode, strconv.Itoa(res.StatusCode))
+		// treat 4XX and 5XX as errors for a client
+		if res.StatusCode >= 400 && res.StatusCode < 600 {
+			span.SetTag(ext.Error, fmt.Errorf("%s", res.Status))
+		}
+	}
+	return res, err
+}
+
+// WrapServer wraps a http.Handler to add distributed tracing to a Twirp server.
+func WrapServer(h http.Handler, opts ...Option) http.Handler {
+	cfg := new(config)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		opts := []tracer.StartSpanOption{
+			tracer.SpanType(ext.SpanTypeWeb),
+			tracer.Tag(ext.HTTPMethod, r.Method),
+			tracer.Tag(ext.HTTPURL, r.URL.Path),
+		}
+		if !math.IsNaN(cfg.analyticsRate) {
+			opts = append(opts, tracer.Tag(ext.EventSampleRate, cfg.analyticsRate))
+		}
+		if spanctx, err := tracer.Extract(tracer.HTTPHeadersCarrier(r.Header)); err == nil {
+			opts = append(opts, tracer.ChildOf(spanctx))
+		}
+
+		span, ctx := tracer.StartSpanFromContext(r.Context(), "twirp.request", opts...)
+		defer span.Finish()
+
+		ctx = context.WithValue(ctx, httpSpanKey, span)
+		r = r.WithContext(ctx)
+		h.ServeHTTP(w, r)
+	})
+}
+
+// NewServerHooks creates the callback hooks for a twirp server to perform tracing.
+func NewServerHooks(opts ...Option) *twirp.ServerHooks {
+	cfg := new(config)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+	return &twirp.ServerHooks{
+		RequestReceived:  requestReceivedHook(cfg),
+		RequestRouted:    requestRoutedHook(cfg),
+		ResponsePrepared: responsePreparedHook(cfg),
+		ResponseSent:     responseSentHook(cfg),
+		Error:            errorHook(cfg),
+	}
+}
+
+func requestReceivedHook(cfg *config) func(context.Context) (context.Context, error) {
+	return func(ctx context.Context) (context.Context, error) {
+		pkg, ok := twirp.PackageName(ctx)
+		if !ok {
+			pkg = "unknown_package"
+		}
+		svc, ok := twirp.ServiceName(ctx)
+		if !ok {
+			svc = "unknown_service"
+		}
+		name := fmt.Sprintf("%s.%s server", pkg, svc)
+		if cfg.serviceName != "" {
+			name = cfg.serviceName
+		}
```

Comment:
We can improve this part. A few things:

1. This seems to be the same in both places. Let's factor it out into a separate function that takes a context and a default, and improve on it.
2. Why are we calling all those methods and (the potentially expensive) `fmt.Sprintf` if in the end we might just cancel everything out by replacing from config?
3. We can better take advantage of unavailable information and create better fallbacks than for example `unknown_package.hat server`.
4. It's quite unusual to have spaces in a service name. Have we seen that anywhere else?

Here's an example:

**In this function**:
```go
name := cfg.serviceName
if name == "" {
    name = obtainServiceName(ctx)
}
```

**Later on**:
```go
func obtainServiceName(ctx context.Context) string {
    pkg, okpkg := twirp.PackageName(ctx)
    svc, oksvc := twirp.ServiceName(ctx)
    switch {
    case okpkg, oksvc:
        return fmt.Sprintf("%s.%s-server", pkg, svc)
    case okpkg, !oksvc:
        return fmt.Sprintf("%s.twirp-server", pkg)
    case !okpkg, oksvc:
        return fmt.Sprintf("%s.%s-server", os.Args[0], svc)
    default:
        return fmt.Sprintf("%s.twirp-server", os.Args[0])
    }
}
```
It doesn't have to be like my example, but I mean to cover all cases and be a bit nicer. WDYT?

Reply:
After some more thought, I think my idea to try and derive the service name from the package/service/method being called isn't good and these should just be tags on the span.
Eg: tag them as `twirp.package`, `twirp.service`, and `twirp.method`. This will render as a group.

Reply:
Yeah that sounds about right 👍